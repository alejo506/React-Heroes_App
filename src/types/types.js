
//Este objeto va a contener todos los types de mi aplicaci√≥n sin importar cuantos sean
//Este objeto lo utilizamos en auth
export const types = {
    login: '[auth] login',
    logout: '[auth] logout',

}