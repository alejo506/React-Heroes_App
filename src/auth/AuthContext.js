//Hacemos la importación de Context
import { createContext } from 'react';

//Este es nuestro context. Lo utilizaremos en HeroesApp.js, se puede utilizar en el index.js pero es mejor dejar nuestro index.js lo más limpio posible, lo hacemos el HeroesApp porque ahí es donde ya comienza a desplegarse toda mi aplicación
export const AuthContext = createContext();


