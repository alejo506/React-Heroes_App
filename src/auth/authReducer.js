import { types } from "../types/types";

// //Este será el estado si nuestro usuario está autenticado, y si no está autenticado no va a tener la propiedad "name" y el logged estará en "false"
// const state = {
//     name: 'Alejandro',
//     logged: true
// }
//Exportamos el reducer
export const authReducer = ( state = {}, action)=>{//El reducer recibe el state elcual es un objeto. La action es la accion que vamos a realizar

    
    
    switch ( action.type ) {//Evaluaremos el action.type
        //Las dos acciones que vamos a realizar es el Login y el Logout
        case types.login:
            
        return{
            //Retornamos todo lo que viene en el action.payload (...action.payload)
            ...action.payload, //Tomamos lo que viene en el payload de la accion y se establece abajo el logged en true, esto es porque si enviamos un name:'Alejandro' este va a ser el payload de la de accion y si recibe dicha propiedad entonces el logged cambia a true por el contrario, si no enviamos el name:'Alejandro' solo se establece la propiedad logged en false(en este caso es logout)
            logged: true //La autenticación del usuario. Si pasa la autenticación va a ser tru
        }
        
        case types.logout:
            
            return { //En el logout no nos interesará nada del payload, solo retornaremos el estamos de la propiedad logged en "false"
                logged: false
            }
            
            default:
                return state; //Retornamos el estado tal y como se encuentra originalmente
            }
            
}