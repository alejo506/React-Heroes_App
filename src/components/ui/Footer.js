import React from 'react'

export const Footer = () => {
    return (
<footer className="bg-dark text-center text-white" style={{bottom:0, marginTop:233}}>

  <div className="text-center p-3" style={{backgroundColor:"#000000", opacity:0.2}}>
    Â© 2021 Copyright:
    <a className="text-white ms-1" href="https://www.youtube.com/watch?v=LXb3EKWsInQ"> Alejandro CM</a>
  </div>
</footer>
    )
}
