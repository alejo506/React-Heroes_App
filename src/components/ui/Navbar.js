import React, { useContext } from 'react'
import { Link, NavLink, useHistory } from 'react-router-dom'
import { AuthContext } from '../../auth/AuthContext';
import { types } from '../../types/types';

export const Navbar = () => { //En Navbar no se puede utilizar el {history} ya que history en Navbar tiene el valor de "undefined" es to es debido a que Navbar no se encuentra dentro de las rutas(Route), Navbar como tal es un componente que está por aparte, para hacer uso de rutas tengo que hacer uso obligatorio del customHook useHistory. Otra forma de hacer, pero, no es lo mejor, ya que, es para evitar escribir menos codigo, sería enviar el history al Navbar desde el componente de ruta DashboarRoutes

    // const [show, setShow] = React.useState(false);

    //Con user:{name} extraemos la propiedad "name" del objeto "user"
    const { user: { name }, dispatch } = useContext(AuthContext);
    
    
    const {replace} =useHistory();

    //Hacemos el Logout
    const handleLogout= ()=>{
        
        console.log('Click!');//Video 204.Cuando simulamos el evento click al hacer las pruebas del componente Nvbar, utilizamos este console para ver la simulacion del onClick
        
        replace('/login');

        dispatch({
            type: types.logout
        });


    }

    return (

        <nav className="navbar navbar-expand-lg navbar-dark bg-dark">
            <div className="container-fluid">
                <div className="nav-item ">
                    <Link
                        className="navbar-brand nav-link"
                        to="/"
                        style={{ marginLeft: 15 }}
                    >
                        Asociaciones
                    </Link>
                </div>
                <button className="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
                    <span className="navbar-toggler-icon"></span>
                </button>
                <div className="collapse navbar-collapse" id="navbarSupportedContent">
                    <ul className="navbar-nav me-auto mb-1 mb-lg-0">
                        <div className="nav-item ">
                            <NavLink
                                activeClassName="active"
                                className="nav-link"
                                exact
                                to="/marvel"
                            >
                                Marvel
                    </NavLink>
                        </div>

                        <div className="nav-item ">

                            <NavLink
                                activeClassName="active"
                                className="nav-link"
                                exact
                                to="/dc"
                            >
                                DC
                    </NavLink>
                        </div>

                        <div className="nav-item">

                            <NavLink
                                activeClassName="active"
                                className="nav-link"
                                exact
                                to="/search"
                            >
                                Search
                    </NavLink>
                        </div>
                    </ul>

                    <div className="nav-item dropdown ">
                        <a className="navbar-brand nav-link dropdown-toggle" id="navbarDarkDropdownMenuLink" role="button" data-bs-toggle="dropdown" aria-expanded="false">
                            <lord-icon
                                src="https://cdn.lordicon.com/dxjqoygy.json"
                                trigger="hover"
                                colors="primary:#e4e4e4,secondary:#e4e4e4"
                                stroke="60"
                                style={{ width: 38, height: 38 }}>
                            </lord-icon>
                            {/* {name} */} {/*Esto es para agregar el nombre del user logueado a la par del iconUser*/}
                        </a>
                        <ul className="dropdown-menu dropdown-menu-dark" aria-labelledby="navbarDarkDropdownMenuLink">

                            <span className="ms-3 textInfo" style={{color:"#5cb85c"}}> {name} {/*Esto es para agregar el nombre del user logueado a la par del iconUser*/}</span>
                            <button
                                
                                className="dropdown-item"
                                onClick ={handleLogout}
                            >
                                Logout
                            </button>

                        </ul>
                    </div>
                </div>
            </div>
        </nav>
    )
}