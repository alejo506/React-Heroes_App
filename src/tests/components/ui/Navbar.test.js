import React from 'react';
import { mount } from "enzyme/build";
//Como queremos limpiar el Mock en historyMock importamos la sigueinte librería @testing-library/jest-dom
import '@testing-library/jest-dom';
import { Router,HashRouter, MemoryRouter } from "react-router-dom";
import { AuthContext } from "../../../auth/AuthContext";
import { Navbar } from "../../../components/ui/Navbar";
import { types } from '../../../types/types';

describe('pruebas en el comoponente <Navbar />', () => {

    //-----------------------------------------------------------
    //Para hacer la prueba con el history. El historyMock se lo podemos enviar al Router y nuestro componente cuando va a usar el useHistory entonces va a ver y va a utilizar el history que tiene ese router(Ojo recordemos que en lugar del router estamos utilizando el HashRouter, pero como el HashRouter ignora el historial, obligatoriamente tengo que utilizar el Router para hacer la prueba)
    const historyMock = { //Podemos agregar las propiedades que queramos
        push: jest.fn(),
        replace: jest.fn(),
        location: {},
        listen: jest.fn(), //Al profe le dio un error que dice que history.listen not is a function, entonces lo agregó y simuló la función, ya que listen es una funcion
        createHref: jest.fn()  //createHref este igual dice que no es una función history.createHref is not a function, lo agregamos y simulamos la función para quitar el error
    }

    //-----------------------------------------------------------

    const contextValue = {
        dispatch: jest.fn(),
        user: {
            logged: true,
            name: 'Valentina'
        }
    }
    //Como tenemos que utilizar Context y evaluar los hijos de un higher order component utilizamos el mount
    const wrapper = mount(
        <AuthContext.Provider value={contextValue}>
            {/* Error al no usar el MemoryRouter. Error: Uncaught [TypeError: Cannot destructure property 'replace' of '(0 , _reactRouterDom.useHistory)(...)' as it is undefined.] */}
            <MemoryRouter>
                {/* Para la prueba tengo que utilizar Router porque HasRouter ignora el historial . Advertencia: <HashRouter> ignora la propiedad del historial. Para usar un historial personalizado, use `import {Router}` en lugar de `import {HashRouter as Router}`.*/}
                <Router history={historyMock}>
                    <Navbar />
                </Router>
            </MemoryRouter>
        </AuthContext.Provider>
    );

    //Limpiar los Mocks. Siempre que hagamos algun mock es buena practica limpiarlo
    afterEach(() => {
        jest.clearAllMocks();
    })

    test('debe de mostrarse correctamente', () => {

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.textInfo').text().trim()).toBe('Valentina');
    });

    test('debe de llamar el Logout y usar el history ', () => {

    //-----------------------------------------------------------
        // En la funcion handleLogout desde nuestro componente Navbar agregamos un console.log('Click') para que se imprima y simule el evento onClick
        wrapper.find('.dropdown-item').prop('onClick')();

        const type = types.logout;

        expect(contextValue.dispatch).toHaveBeenCalledWith({ type });
    //-----------------------------------------------------------

    //history. La prueba del history sirve por si alguien quitara el history.replace('/login') de la función handleLogout desde el componente Navbar.js, entonces dará un error, eso quiere decir que va a necesitar de ese history para que funcione
    expect(historyMock.replace).toHaveBeenCalledWith('/login');

    })


});
