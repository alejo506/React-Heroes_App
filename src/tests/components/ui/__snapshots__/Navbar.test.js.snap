// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en el comoponente <Navbar /> debe de mostrarse correctamente 1`] = `
<nav
  className="navbar navbar-expand-lg navbar-dark bg-dark"
>
  <div
    className="container-fluid"
  >
    <div
      className="nav-item "
    >
      <a
        className="navbar-brand nav-link"
        onClick={[Function]}
        style={
          Object {
            "marginLeft": 15,
          }
        }
      >
        Asociaciones
      </a>
    </div>
    <button
      aria-controls="navbarSupportedContent"
      aria-expanded="false"
      aria-label="Toggle navigation"
      className="navbar-toggler"
      data-bs-target="#navbarSupportedContent"
      data-bs-toggle="collapse"
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      className="collapse navbar-collapse"
      id="navbarSupportedContent"
    >
      <ul
        className="navbar-nav me-auto mb-1 mb-lg-0"
      >
        <div
          className="nav-item "
        >
          <a
            aria-current={null}
            className="nav-link"
            onClick={[Function]}
          >
            Marvel
          </a>
        </div>
        <div
          className="nav-item "
        >
          <a
            aria-current={null}
            className="nav-link"
            onClick={[Function]}
          >
            DC
          </a>
        </div>
        <div
          className="nav-item"
        >
          <a
            aria-current={null}
            className="nav-link"
            onClick={[Function]}
          >
            Search
          </a>
        </div>
      </ul>
      <div
        className="nav-item dropdown "
      >
        <a
          aria-expanded="false"
          className="navbar-brand nav-link dropdown-toggle"
          data-bs-toggle="dropdown"
          id="navbarDarkDropdownMenuLink"
          role="button"
        >
          <lord-icon
            colors="primary:#e4e4e4,secondary:#e4e4e4"
            src="https://cdn.lordicon.com/dxjqoygy.json"
            stroke="60"
            style={
              Object {
                "height": 38,
                "width": 38,
              }
            }
            trigger="hover"
          />
           
        </a>
        <ul
          aria-labelledby="navbarDarkDropdownMenuLink"
          className="dropdown-menu dropdown-menu-dark"
        >
          <span
            className="ms-3 textInfo"
            style={
              Object {
                "color": "#5cb85c",
              }
            }
          >
             
            Valentina
             
          </span>
          <button
            className="dropdown-item"
            onClick={[Function]}
          >
            Logout
          </button>
        </ul>
      </div>
    </div>
  </div>
</nav>
`;
